llama_index/chat_engine/types.py

```diff
@@ -195,6 +197,183 @@ async def aprint_response_stream(self) -> None:
             print(token, end="", flush=True)
 
 
+@dataclass
+class CohereAgentChatResponse(AgentChatResponse):
+    """Cohere Agent chat response. Adds citations and documents to the response."""
+
+    citations: List[dict] = field(default_factory=list)
+    documents: List[dict] = field(default_factory=list)
+
+
+@dataclass
+class CohereStreamingAgentChatResponse(StreamingAgentChatResponse):
+    """Streaming chat response to user and writing to chat history."""
+
+    citations: List[dict] = field(default_factory=list)
+    documents: List[dict] = field(default_factory=list)
+    _documents_queue: queue.Queue = field(default_factory=queue.Queue)
+    _documents_aqueue: asyncio.Queue = field(default_factory=asyncio.Queue)
+    _citations_queue: queue.Queue = field(default_factory=queue.Queue)
+    _citations_aqueue: asyncio.Queue = field(default_factory=asyncio.Queue)
+
+    def put_in_the_queue(self, queue: queue.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+
+    def aput_in_the_queue(self, queue: asyncio.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+        self._new_item_event.set()
+
+    def write_response_to_history(
+        self, memory: BaseMemory, raise_error: bool = False
+    ) -> None:
+        if self.chat_stream is None:
+            raise ValueError(
+                "chat_stream is None. Cannot write to history without chat_stream."
+            )
+
+        # try/except to prevent hanging on error
+        try:
+            final_text = ""
+            for chat in self.chat_stream:
+                # Cohere response queue
+                self.put_in_queue(chat.delta)
+                if chat.raw is not None:
+                    # Citations queue
+                    if (
+                        chat.raw.get("event_type", "")
```

Comment:
you can clean up the .get() here by assigning to a variable before the if statements, `event_type = chat.raw.get("event_type", "")`

Reply:
No, the chat.raw can be None, and None doesn't provide the get method.